File: Hashing/Group Anagrams/group_anagrams.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> map;

    for (const string& s : strs) {
        string key = s;
        sort(key.begin(), key.end());
        map[key].push_back(s);
    }

    vector<vector<string>> result;
    for (const auto& entry : map) {
        result.push_back(entry.second);
    }

    return result;
}

int main() {
    int n;
    cout << "Size of array: ";
    cin >> n;

    vector<string> strs(n);
    cout << "Enter space-separated strings: ";
    for (int i = 0; i < n; ++i) {
        cin >> strs[i];
    }

    vector<vector<string>> grouped = groupAnagrams(strs);
    cout << "Grouped Anagrams:\n";
    for (const auto& group : grouped) {
        cout << "[ ";
        for (const string& word : group) {
            cout << word << " ";
        }
<<<<<<< HEAD
        cout << "]\n";
=======
        cout << "]";
>>>>>>> 25bcd466af18b3b64323a03cbd2d20735bd0c30b
    }

    return 0;
}

File: Hashing/Group Anagrams/GroupAnagrams.java
import java.util.*;

public class GroupAnagrams {
    public static List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();

        for (String s : strs) {
            char[] chars = s.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);

            map.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }

        return new ArrayList<>(map.values());
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Size of array: ");
        int n = sc.nextInt();
        sc.nextLine(); 

        System.out.print("Enter space-separated strings: ");
        String[] strs = sc.nextLine().split(" ");

        if (strs.length != n) {
            System.out.println(" Expected " + n + " strings, but got " + strs.length);
        } else {
            List<List<String>> result = groupAnagrams(strs);
            System.out.println("Grouped Anagrams:");
            for (List<String> group : result) {
                System.out.println(group);
            }
        }
    }
}